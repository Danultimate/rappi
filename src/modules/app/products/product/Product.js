import React from 'react';
import './Product.css';

const Product = ({
  quantity,
  price,
  name,
  available
}) => (
  <tr>
    <td>{name}</td>
    <td>{price}</td>
    <td>{quantity}</td>
    <td>{available ? null : 'NA'}</td>
  </tr>
)

export default Product;