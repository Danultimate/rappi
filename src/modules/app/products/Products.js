import React from 'react';
import Product from './product/Product';
import { formatPrice } from '../utils/numbers';

const filterProducts = (products, filters) => {
  const {
    sublevels,
    available,
    minPrice,
    maxPrice,
    minStock,
    maxStock
  } = filters;

  return products.filter(product => {
    return (!sublevels.length || sublevels.includes(product.sublevel_id))
      && (!available || product.available === available)
      && (!minPrice || formatPrice(product.price) >= minPrice)
      && (!maxPrice || formatPrice(product.price) <= maxPrice)
      && (!minStock || product.quantity >= minStock)
      && (!maxStock || product.quantity <= minStock)
  });
}

// Just by ascending order, for example purpose
const orderProducts = (products, priority) => {
  priority.map(criteria => products.sort((a, b) => {
    if (criteria === 'price') {
      return formatPrice(a[criteria]) - formatPrice(b[criteria])
    }
    return a[criteria] - b[criteria]
  }));
  return products;
}

const hasOrder = (criteria, priority) => {
  return priority.includes(criteria) ? '*' : null;
}

const Products = ({ products, filters, priority, handleChange }) => {
  const filteredProducts = filterProducts(products, filters);
  const orderedFilteredProducts = orderProducts(filteredProducts, priority);

  return (
    <table className="table">
      <thead>
        <tr>
          <th scope="col">Name</th>
          <th scope="col" onClick={() => handleChange('price')}>
            Price {hasOrder('price', priority)}
          </th>
          <th scope="col" onClick={() => handleChange('quantity')}>
            Quantity {hasOrder('quantity', priority)}
          </th>
          <th scope="col" onClick={() => handleChange('available')}>
            NA {hasOrder('available', priority)}
          </th>
        </tr>
      </thead>
      <tbody>
        {orderedFilteredProducts.map((product) => (
          <Product key={product.id} {...product} />
        ))}
      </tbody>
    </table>
  )
};

export default Products;