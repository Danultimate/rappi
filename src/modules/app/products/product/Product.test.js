import React from 'react';
import Product from './Product';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import products from '../../store/products.json';

describe('Product', () => {
  configure({ adapter: new Adapter() });

  it('renders properly', () => {
    const props = { ...products[0] };
    const wrapper = shallow(<Product {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
