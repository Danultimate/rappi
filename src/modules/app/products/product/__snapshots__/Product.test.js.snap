// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders properly 1`] = `ShallowWrapper {}`;
